pennylane/capture/base_interpreter.py

```diff
@@ -0,0 +1,415 @@
+            Calling the returned decorator with a function will place the function into the
+            primitive registrations map.
+
+        .. code-block:: python
+
+            my_primitive = jax.core.Primitive("my_primitve")
+
+            @Interpreter_Type.register(my_primitive)
+            def handle_my_primitive(self: Interpreter_Type, *invals, **params)
+                return invals[0] + invals[1] # some sort of custom handling
+
+        """
+
+        def decorator(f: Callable) -> Callable:
+            cls._primitive_registrations[primitive] = f
+            return f
+
+        return decorator
+
+    # pylint: disable=unidiomatic-typecheck
+    def read(self, var):
+        """Extract the value corresponding to a variable."""
+        if self._env is None:
+            raise ValueError("_env not yet initialized.")
+        if type(var) is jax.core.Literal:
+            return var.val
+        return self._op_math_cache.get(var, self._env[var])
+
+    def setup(self) -> None:
+        """Initialize the instance before interpretting equations.
+
+        Blank by default, this method can initialize any additional instance variables
+        needed by an interpreter. For example, a device interpreter could initialize a statevector,
+        or a compilation interpreter could initialize a staging area for the latest operation on each wire.
+
+        """
+
+    def cleanup(self) -> None:
+        """Perform any final steps after iterating through all equations.
+
+        Blank by default, this method can clean up instance variables. Particularily,
+        this method can be used to deallocate qubits and registers when converting to
+        catalyst variant jaxpr.
+        """
+
+    def interpret_operation(self, op: "pennylane.operation.Operator"):
+        """Interpret a PennyLane operation instance.
+
+        Args:
+            op (Operator): a pennylane operator instance
+
+        Returns:
+            Any
+
+        This method is only called when the operator's output is a dropped variable,
+        so the output will not effect later equations in the circuit.
+
+        See also: :meth:`~.interpret_operation_eqn`.
+
+        """
+        return op._unflatten(*op._flatten())  # pylint: disable=protected-access
+
+    def interpret_operation_eqn(self, eqn: "jax.core.JaxprEqn"):
+        """Interpret an equation corresponding to an operator.
+
+        Args:
+            eqn (jax.core.JaxprEqn): a jax equation for an operator.
+
+        See also: :meth:`~.interpret_operation`.
+
+        """
+
+        invals = (self.read(invar) for invar in eqn.invars)
+        with qml.QueuingManager.stop_recording():
+            op = eqn.primitive.impl(*invals, **eqn.params)
+        if isinstance(eqn.outvars[0], jax.core.DropVar):
+            return self.interpret_operation(op)
+
+        self._op_math_cache[eqn.outvars[0]] = op
+        return op
+
+    def interpret_measurement_eqn(self, primitive, *invals, **params):
```

Comment:
Is there a reason that `interpret_measurement_eqn` does not take an `eqn` like `interpret_operation_eqn`? Would be neat to have the same signature for both, and seems possible from looking at the `eval` code :)

Reply:
So for operators, we often need to know whether or not it's a dropped variable. So we need to also include the outval.  I considered doing something like:
```
def interpreter_operator_eqn(self, primitive, outvar, *invals, **params):
```
but then it is basically just the eqn. 

But generally I kinda prefer having a syntax for handling something like `(*invals, **params)`, because then it looks like the call signature for the primitive itself.

Reply:
Thanks, this makes sense! I like your reasoning for the signature to match that of the primitive itself!

I don't want to push on the details too much, because I agree with you that we mostly want to cover a big part of the standard interpreters we will want to build.
For completeness, and to have it discussed, I'd leave the following thoughts:
- The current `interpret_operator_eqn` puts a precedence for an interpreting function needing to know the some extra info from the `eqn`. In case we ever want an interpreter with `interpret_measurement_eqn` that is like that, we would have to write a new `eval` function for the interpreter, because it will not be easy to change the signature of `interpret_measurement_eqn` in the base class. Passing `eqn` is more flexible in that sense, and leaves all details of handling an equation to the custom `interpret_*_eqn` function rather than a custom `eval` that only swaps out some call signatures.
- Preaching in the exact opposite direction, making `interpret_operator_eqn` explicitly take the `outvar` argument would make the entire setup more self-consistent.
- The current approach nicely takes into account the concrete boundary conditions we have: There currently _is_ a difference between the required information in the interpreting functions, and the code reflects that. Also, where possible, the primitive's signature is reflected.